compiler/circle-opselector/src/OpSelector.cpp

```diff
@@ -0,0 +1,400 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "OpSelector.h"
+
+#include <luci/Service/ChangeOutputs.h>
+
+#include <luci/Import/GraphBuilder.h>
+#include <luci/Import/GraphBuilderContext.h>
+#include <luci/Import/GraphBuilderRegistry.h>
+#include <luci/Import/CircleReader.h>
+#include <luci/Import/Nodes/CircleConst.h>
+
+#include <luci/Profile/CircleNodeID.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+#include <loco/IR/Graph.h>
+#include <luci/IR/CircleNode.h>
+#include <loco/IR/Node.h>
+
+#include <oops/UserExn.h>
+
+#include <cassert>
+#include <iostream>
+#include <map>
+
+#define MAIN_SUBGRAPH 0
+
+// TODO: Re-implement to use just module.
+namespace opselector
+{
+
+void OpSelector::check_connected(std::vector<const luci::CircleNode *> &selected_nodes,
+                                 std::set<uint32_t> &used_output_tensors,
+                                 std::set<uint32_t> &graph_inputs,
+                                 std::set<uint32_t> &graph_outputs)
+{
+  const auto &operators = _reader.operators();
+
+  std::set<uint32_t> selected_input_tensors;
+  std::set<uint32_t> selected_output_tensors;
+
+  std::vector<const circle::OperatorT *> selected_operators;
+
+  // enroll all output nodes.
+  for (auto &op : operators)
+    for (auto output : op.get()->outputs)
+      used_output_tensors.insert(output);
+
+  for (auto input : _reader.inputs()) // graph's input must not have preceding node.
+    used_output_tensors.insert(input);
+
+  // select operators.
+  for (auto cnode : selected_nodes)
+  {
+    uint32_t node_id = luci::get_node_id(cnode);
+    selected_operators.push_back(operators[node_id].get()); // put selected nodes in vector.
+
+    if (cnode->name().find("while") != std::string::npos ||
+        cnode->name().find("if") != std::string::npos) // if has while of if node,
+      _has_subgraph = true; // A flag indicating whether to copy the subgraph or not,
+  }
+
+  print_selected_nodes(selected_nodes);
+
+  // add all selected tensors.
+  for (auto op : selected_operators)
+  {
+    for (auto input : op->inputs)
+      selected_input_tensors.insert(input);
+    for (auto output : op->outputs)
+      selected_output_tensors.insert(output);
+  }
+  // find and add unconnected node's output
+  for (auto op : selected_operators)
+  {
+    bool output_connected = false;
+
+    for (auto output : op->outputs) // check connection
+      if (selected_input_tensors.find(output) != selected_input_tensors.end())
+        output_connected = true;
+
+    if (not output_connected) // if not connected other selected nodes, add all outputs
+      for (auto output : op->outputs)
+        graph_outputs.insert(output);
+  }
```

Comment:
I don't think it will work for some graphs. For example, a node (= operator) has multiple outputs and only one of them is connected with other node. As I mentioned several times, please refer to `select_operator.py`, it is simple and works for general cases also.

Reply:
Good! After you update, the case above seems to be covered. However, it still requires modification. Again, please find the condition from `select_operator.py`.

https://github.com/Samsung/ONE/blob/3e4a55187dbb789173986d526443e086b8d8e158/tools/tflitefile_tool/select_operator.py#L1362-L1363

If the associated buffer length is not zero, it means it is constant tensor, which is not the input of graph.

The author of `select_operator` (@hseok-oh) also removes the case of associated buffer `index = 0`.

Please find the comment from `schema.fbs` for the meaning of `buffer idx = 0`.

```js
table Tensor {
  // The tensor shape. The meaning of each entry is operator-specific but
  // builtin ops use: [batch size, height, width, number of channels] (That's
  // Tensorflow's NHWC).
  shape:[int];
  type:TensorType;
  // An index that refers to the buffers table at the root of the model. Or,
  // if there is no data buffer associated (i.e. intermediate results), then
  // this is 0 (which refers to an always existent empty buffer).
  //
  // The data_buffer itself is an opaque container, with the assumption that the
  // target device is little-endian. In addition, all builtin operators assume
  // the memory is ordered such that if `shape` is [4, 3, 2], then index
  // [i, j, k] maps to data_buffer[i*3*2 + j*2 + k].
  buffer:uint;
```